src/crafting.rs

```diff
@@ -148,6 +150,9 @@ fn calculate_precise_min_crafting_cost(
                     return None;
                 }
             }
+
+            used_recipe(&recipe);
```

Comment:
I think that this marks a recipe as used even when the item to craft has been purchased from the trading post. If that works for you, then maybe just something to note in a comment.

Only adding recipes for items that need crafting would be a bit more complicated, but I think it could be done in a similar way to how `context.purchases` is generated, but just with inverted conditions. Something like:

```
@@ -98,6 +98,7 @@ struct PreciseCraftingCost {
 
 struct PreciseCraftingCostContext {
     purchases: Vec<(u32, Rational32, Source)>,
+    crafted: Vec<u32>,
     crafting_steps: Rational32,
 }
 
@@ -119,6 +120,7 @@ fn calculate_precise_min_crafting_cost(
         Rational32::from(recipe.map(|recipe| recipe.output_item_count).unwrap_or(1));
 
     let purchases_ptr = context.purchases.len();
+    let crafted_ptr = context.crafted.len();
     let crafting_steps_before = context.crafting_steps;
 
     let crafting_cost = recipe.and_then(|recipe| {
@@ -181,6 +183,11 @@ fn calculate_precise_min_crafting_cost(
 
     if source == Source::Crafting {
         context.crafting_steps += item_count / output_item_count;
+        for crafted_item_id in context.crafted.drain(crafted_ptr..) {
+            if let Some(recipe) = recipes_map.get(&crafted_item_id) {
+                used_recipe(recipe);
+            }
+        }
     } else {
         for (purchase_id, purchase_quantity, purchase_source) in
             context.purchases.drain(purchases_ptr..)
@@ -197,6 +204,8 @@ fn calculate_precise_min_crafting_cost(
 
     if source != Source::Crafting {
         context.purchases.push((item_id, item_count, source));
+    } else {
+        context.crafted.push(item_id);
     }
     if source == Source::TradingPost {
         tp_listings_map
```

Reply:
```diff
@@ -98,6 +98,7 @@ struct PreciseCraftingCost {
 
 struct PreciseCraftingCostContext {
     purchases: Vec<(u32, Rational32, Source)>,
+    crafted: Vec<u32>,
     crafting_steps: Rational32,
 }
 
@@ -108,7 +109,6 @@ fn calculate_precise_min_crafting_cost(
     item_count: Rational32,
     recipes_map: &HashMap<u32, Recipe>,
     items_map: &HashMap<u32, api::Item>,
-    used_recipe: &mut dyn FnMut(&Recipe) -> (),
     tp_listings_map: &mut BTreeMap<u32, ItemListings>,
     context: &mut PreciseCraftingCostContext,
     opt: &CraftingOptions,
@@ -119,6 +119,7 @@ fn calculate_precise_min_crafting_cost(
         Rational32::from(recipe.map(|recipe| recipe.output_item_count).unwrap_or(1));
 
     let purchases_ptr = context.purchases.len();
+    let crafted_ptr = context.crafted.len();
     let crafting_steps_before = context.crafting_steps;
 
     let crafting_cost = recipe.and_then(|recipe| {
@@ -135,7 +136,6 @@ fn calculate_precise_min_crafting_cost(
                     ingredient_count,
                     recipes_map,
                     items_map,
-                    used_recipe,
                     tp_listings_map,
                     context,
                     opt,
@@ -151,8 +151,6 @@ fn calculate_precise_min_crafting_cost(
                 }
             }
 
-            used_recipe(&recipe);
-
             Some(cost)
         }
     });
@@ -180,6 +178,7 @@ fn calculate_precise_min_crafting_cost(
     };
 
     if source == Source::Crafting {
+        context.crafted.push(item_id);
         context.crafting_steps += item_count / output_item_count;
     } else {
         for (purchase_id, purchase_quantity, purchase_source) in
@@ -192,6 +191,7 @@ fn calculate_precise_min_crafting_cost(
                     .pending_buy_quantity -= purchase_quantity;
             }
         }
+        context.crafted.drain(crafted_ptr..);
         context.crafting_steps = crafting_steps_before;
     }
 
@@ -278,6 +281,7 @@ pub fn calculate_crafting_profit(
 
         let mut context = PreciseCraftingCostContext {
             purchases: vec![],
+            crafted: vec![],
             crafting_steps: Rational32::zero(),
         };
 
@@ -289,7 +293,6 @@ pub fn calculate_crafting_profit(
             output_item_count.into(),
             recipes_map,
             items_map,
-            &mut used_recipe,
             &mut tp_listings_map,
             &mut context,
             opt,
@@ -321,6 +324,12 @@ pub fn calculate_crafting_profit(
             break;
         }
 
+        for item_id in &context.crafted {
+            if let Some(recipe) = recipes_map.get(item_id) {
+                used_recipe(recipe);
+            }
+        }
+
         for (purchase_id, count, purchase_source) in &context.purchases {
             if *purchase_source != Source::TradingPost {
                 continue;
```

Reply:
Yeah, I moved it down. I originally didn't want to as I figured I'd need to pull recipe out of the scope, and overlooked it or simply noted that every example I'd seen was account bound anyway. Looking through the logic, `unwrap` should be fine if we get to that point.

Reply:
It's a bit hard to follow in my code, but https://github.com/t-mw/gw2-arbitrage/blob/master/src/crafting.rs#L178 doesn't necessarily mean that the item will be crafted if it's an ingredient, because it might be rolled back by a parent recipe in https://github.com/t-mw/gw2-arbitrage/blob/master/src/crafting.rs#L190. I think the changes in my last comment should work for that case, I'll try adding it after you merge.